File: kb_python/query_kb/kb_rpc_client.py
import json
import datetime
import uuid
from datetime import datetime, timezone
class KB_RPC_Client:
    """
    A class to handle the RPC client for the knowledge base.
    """
    def __init__(self, kb_search):
        self.kb_search = kb_search
        self.conn = self.kb_search.conn
        self.cursor = self.kb_search.cursor 
    
    def find_rpc_client_id(self, node_name, properties, node_path):
        """
        Find the node id for a given node name, properties, node path, and data.
        """
        print(node_name, properties, node_path)
        result = self.find_node_ids(node_name, properties, node_path)
        if len(result) == 0:
            raise ValueError(f"No node found matching path parameters: {node_name}, {properties}, {node_path}")
        if len(result) > 1:
            raise ValueError(f"Multiple nodes found matching path parameters: {node_name}, {properties}, {node_path}")
        return result
    
    def find_rpc_client_ids(self, node_name, properties, node_path):
        """
        Find the node id for a given node name, properties, node path :
        """
        print(node_name, properties, node_path)
        self.kb_search.clear_filters()
        self.kb_search.search_label("KB_RPC_CLIENT_FIELD")
        if node_name is not None:
            self.kb_search.search_name(node_name)
        if properties is not None:
            for key in properties:
                self.kb_search.search_property_value(key, properties[key])
        if node_path is not None:
            self.kb_search.search_path(node_path)
        node_ids = self.kb_search.execute_query()
        
        if node_ids is None:
            raise ValueError(f"No node found matching path parameters: {node_name}, {properties}, {node_path}")
        if len(node_ids) == 0:
            raise ValueError(f"No node found matching path parameters: {node_name}, {properties}, {node_path}")
        return node_ids
    
    def find_rpc_client_keys(self, key_data):
       
            
        return_values = []
        for key in key_data:
            
            return_values.append(key[5])
        return return_values    
    
    def find_free_slots(self, client_path):
        """
        Find the number of free slots (records with is_new_result=FALSE) for a given client_path.
        
        Args:
            client_path (str): LTree compatible path for client
            
        Returns:
            int: Number of free slots available for the client_path
            
        Raises:
            Exception: If no records exist for the specified client_path
        """
        try:
            # First, verify that there are records with the given client_path
            self.cursor.execute("""
                SELECT COUNT(*) 
                FROM rpc_client_table.rpc_client_table 
                WHERE client_path = %s
            """, (client_path,))
            
            total_records = self.cursor.fetchone()[0]
            
            if total_records == 0:
                raise Exception(f"No records found for client_path: {client_path}")
            
            # Now, count the number of free slots (where is_new_result is FALSE)
            self.cursor.execute("""
                SELECT COUNT(*) 
                FROM rpc_client_table.rpc_client_table 
                WHERE client_path = %s AND is_new_result = FALSE
            """, (client_path,))
            
            free_slots = self.cursor.fetchone()[0]
            
            return free_slots
            
        except psycopg2.Error as e:
            # Handle database errors
            raise Exception(f"Database error when finding free slots: {str(e)}")
        
    def find_queued_slots(self, client_path):
        """
        Find the number of queued slots (records with is_new_result=FALSE) for a given client_path.
        
        Args:
            client_path (str): LTree compatible path for client
            
        Returns:
            int: Number of queued slots available for the client_path
            
        Raises:
            Exception: If no records exist for the specified client_path
        """
        try:
            # First, verify that there are records with the given client_path
            self.cursor.execute("""
                SELECT COUNT(*) 
                FROM rpc_client_table.rpc_client_table 
                WHERE client_path = %s
            """, (client_path,))
            
            total_records = self.cursor.fetchone()[0]
            
            if total_records == 0:
                raise Exception(f"No records found for client_path: {client_path}")
            
            # Count the number of queued slots (where is_new_result is FALSE)
            self.cursor.execute("""
                SELECT COUNT(*) 
                FROM rpc_client_table.rpc_client_table 
                WHERE client_path = %s AND is_new_result = TRUE
            """, (client_path,))
            
            queued_slots = self.cursor.fetchone()[0]
            
            return queued_slots
            
        except psycopg2.Error as e:
            # Handle database errors
            raise Exception(f"Database error when finding queued slots: {str(e)}")
    
    def peak_reply_data(self, client_path):
        """
        For a given client_path, find the first record where is_new_result is TRUE
        with the earliest response_timestamp.
        
        Args:
            client_path (str): LTree compatible path for client
            
        Returns:
            tuple: (id, data_dict) where:
                - id is the record ID
                - data_dict is a dictionary containing all fields of the record
                
        Raises:
            Exception: If no records with is_new_result=TRUE are found for the client_path
        """
        try:
            query = """
                SELECT id, request_id, client_path, server_path, transaction_tag, rpc_action,
                       response_payload, response_timestamp, is_new_result, error_message
                FROM rpc_client_table.rpc_client_table
                WHERE client_path = %s AND is_new_result = TRUE
                ORDER BY response_timestamp ASC
                LIMIT 1
            """
            
            # Execute the query
            self.cursor.execute(query, (client_path,))
            
            # Fetch the result
            record = self.cursor.fetchone()
            
            # Check if a record was found
            if record is None:
                return None, None
            
            # Create a dictionary of field names and values
            column_names = [desc[0] for desc in self.cursor.description]
            data_dict = dict(zip(column_names, record))
            
            # Convert UUID to string
            if 'request_id' in data_dict and data_dict['request_id'] is not None:
                data_dict['request_id'] = str(data_dict['request_id'])
            
            # Convert datetime to ISO format string
            if 'response_timestamp' in data_dict and isinstance(data_dict['response_timestamp'], datetime):
                data_dict['response_timestamp'] = data_dict['response_timestamp'].isoformat()
            
            # Get the record ID
            record_id = data_dict.pop('id')
            
            return record_id, data_dict
            
        except psycopg2.Error as e:
            # Handle database errors
            raise Exception(f"Database error when peeking reply data: {str(e)}")


    
    def release_rpc_data(self, client_path, id, max_retries=3, retry_delay=1):
        """
        For a given ID, verify the record matches the client_path and is_new_result is TRUE,
        then set is_new_result to FALSE. Contains protection against parallel transactions.
        
        Args:
            client_path (str): LTree compatible path for client
            id (int): The ID of the record to release
            max_retries (int): Maximum number of retries in case of lock conflicts
            retry_delay (float): Delay in seconds between retries
            
        Returns:
            bool: True if a record was successfully released, False if no matching record was found
            
        Raises:
            Exception: If database error occurs or cannot obtain lock after retries
        """
        retries = 0
        
        while retries <= max_retries:
            try:
                # Create a new cursor for this attempt to ensure clean transaction state
                with self.conn.cursor() as transaction_cursor:
                    # Begin a transaction
                    self.conn.autocommit = False
                    
                    # First, verify the record exists with the given ID, client_path, and is_new_result=TRUE
                    # Use NOWAIT to immediately get an error if the row is locked
                    verify_query = """
                        SELECT id
                        FROM rpc_client_table.rpc_client_table
                        WHERE id = %s AND client_path = %s AND is_new_result = TRUE
                        FOR UPDATE NOWAIT
                    """
                    
                    transaction_cursor.execute(verify_query, (id, client_path))
                    
                    # Check if a record was found
                    if transaction_cursor.fetchone() is None:
                        # No matching record found
                        self.conn.rollback()
                        return False
                    
                    # Update the record
                    update_query = """
                        UPDATE rpc_client_table.rpc_client_table
                        SET is_new_result = FALSE
                        WHERE id = %s
                    """
                    
                    transaction_cursor.execute(update_query, (id,))
                    
                    # Commit the transaction
                    self.conn.commit()
                    
                    # Return success
                    return True
                    
            except psycopg2.errors.LockNotAvailable:
                # This specific exception means another transaction has locked the row
                self.conn.rollback()
                
                if retries < max_retries:
                    # Wait before retrying
                    time.sleep(retry_delay)
                    retries += 1
                else:
                    # Max retries exceeded
                    raise Exception(f"Could not obtain lock on record ID {id} after {max_retries} attempts")
                    
            except psycopg2.Error as e:
                # Handle other database errors
                self.conn.rollback()
                raise Exception(f"Database error when releasing RPC data: {str(e)}")
                
            finally:
                # Reset autocommit to its default state
                self.conn.autocommit = True
        
        # This should not be reached due to the exception in the retry loop,
        # but added as an extra safety measure
        raise Exception(f"Failed to release RPC data for ID {id} after {max_retries} attempts")    


    def clear_reply_queue(self):
        """
        Clear the reply queue by resetting all records in the table.
        For each record:
        - Sets a unique UUID for request_id
        - Sets server_path equal to client_path
        - Resets response_payload to empty JSON object
        - Updates response_timestamp to current UTC time
        - Sets is_new_result to FALSE
        - Clears error_message
        
        Returns:
            int: Number of records updated
        """
        try:
            # Create the update query
            query = """
                UPDATE rpc_client_table.rpc_client_table
                SET request_id = gen_random_uuid(),  -- PostgreSQL function to generate unique UUIDs
                    server_path = client_path,
                    response_payload = '{}',
                    response_timestamp = NOW(),  -- Current UTC timestamp
                    is_new_result = FALSE,
                    error_message = ''
            """
            
            # Execute the query
            self.cursor.execute(query)
            
            # Get the number of affected rows
            row_count = self.cursor.rowcount
            
            # Commit the transaction
            self.conn.commit()
            
            return row_count
            
        except psycopg2.Error as e:
            # Rollback in case of error
            self.conn.rollback()
            raise Exception(f"Failed to clear reply queue: {str(e)}") 
           
    def push_reply_data(self, client_path, request_uuid, server_path, rpc_action, transaction_tag, reply_data, error_text):
        """
        Update the earliest record with is_new_result=False that matches the client_path
        
        Args:
            client_path (str): LTree compatible path for client
            request_uuid (str): UUID of the request
            server_path (str): LTree compatible path for server
            rpc_action (str): Action to be performed
            transaction_tag (str): Transaction tag
            reply_data (dict): Dictionary containing reply data
            error_text (str): Error message if any
            
        Raises:
            Exception: If no available record with is_new_result=False is found
        """
        try:
            # Convert request_uuid to UUID object if it's a string
            if isinstance(request_uuid, str):
                request_uuid = uuid.UUID(request_uuid)
                
            # Start a transaction
            with self.conn.cursor() as cursor:
                # Select for update to lock the row
                cursor.execute("""
                    SELECT id 
                    FROM rpc_client_table.rpc_client_table
                    WHERE client_path = %s AND is_new_result = FALSE
                    ORDER BY response_timestamp ASC
                    LIMIT 1
                    FOR UPDATE NOWAIT
                """, (client_path,))
                
                result = cursor.fetchone()
                if not result:
                    raise Exception("Table_full_exception: No available record with is_new_result=False")
                
                record_id = result[0]
                
                # Update the record
                cursor.execute("""
                    UPDATE rpc_client_table.rpc_client_table
                    SET request_id = %s,
                        server_path = %s,
                        rpc_action = %s,
                        transaction_tag = %s,
                        response_payload = %s,
                        response_timestamp = %s,
                        is_new_result = TRUE,
                        error_message = %s
                    WHERE id = %s
                """, (
                    request_uuid,
                    server_path,
                    rpc_action,
                    transaction_tag,
                    json.dumps(reply_data),
                    datetime.now(timezone.utc),
                    error_text,
                    record_id
                ))
                
                # Commit the transaction
                self.conn.commit()
                
        except psycopg2.Error as e:
            self.conn.rollback()
            if "could not obtain lock" in str(e):
                raise Exception("Parallel operation detected, try again later")
            else:
                raise e
    
    
    
    def list_waiting_jobs(self, client_path=None):
        """
        List all rows in the table where is_new_result is TRUE and optionally 
        matching a specific client_path.
        
        Args:
            client_path (str, optional): If provided, filter results to this client_path
        
        Returns:
            list: A list of dictionaries, each containing the data for one waiting job
                  Each dictionary has keys corresponding to table column names
                  
        Raises:
            Exception: If a database error occurs
        """
        try:
            if client_path is None:
                # No client_path filter, return all waiting jobs
                query = """
                    SELECT id, request_id, client_path, server_path, 
                           response_payload, response_timestamp, is_new_result, error_message
                    FROM rpc_client_table.rpc_client_table
                    WHERE is_new_result = TRUE
                    ORDER BY response_timestamp ASC
                """
                params = ()
            else:
                # Filter by client_path
                query = """
                    SELECT id, request_id, client_path, server_path, 
                           response_payload, response_timestamp, is_new_result, error_message
                    FROM rpc_client_table.rpc_client_table
                    WHERE is_new_result = TRUE AND client_path = %s
                    ORDER BY response_timestamp ASC
                """
                params = (client_path,)
            
            # Execute the query
            self.cursor.execute(query, params)
            
            # Get column names from cursor description
            column_names = [desc[0] for desc in self.cursor.description]
            
            # Fetch all matching records
            records = self.cursor.fetchall()
            
            # Convert to list of dictionaries
            result = []
            for record in records:
                # Create a dictionary for this record
                record_dict = dict(zip(column_names, record))
                
                # Convert UUID to string
                if 'request_id' in record_dict and record_dict['request_id'] is not None:
                    record_dict['request_id'] = str(record_dict['request_id'])
                
                # Convert datetime to ISO format string
                if 'response_timestamp' in record_dict and isinstance(record_dict['response_timestamp'], datetime):
                    record_dict['response_timestamp'] = record_dict['response_timestamp'].isoformat()
                
                # Convert ltree to string
                if 'client_path' in record_dict and record_dict['client_path'] is not None:
                    record_dict['client_path'] = str(record_dict['client_path'])
                
                if 'server_path' in record_dict and record_dict['server_path'] is not None:
                    record_dict['server_path'] = str(record_dict['server_path'])
                
                # Add to result list
                result.append(record_dict)
            
            return result
            
        except psycopg2.Error as e:
            # Handle database errors
            raise Exception(f"Database error when listing waiting jobs: {str(e)}")

File: kb_modules/kb_dotnet/postgres/data_structures/KnowledgeBase/KBRpcClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Npgsql;
using Newtonsoft.Json;

public class KBRpcClient
{
    private readonly KBSearch _kbSearch;
    private readonly string _baseTable;

    public KBRpcClient(KBSearch kbSearch, string database)
    {
        _kbSearch = kbSearch;
        _baseTable = $"{database}_rpc_client";
    }

    private NpgsqlConnection GetConnection()
    {
        return _kbSearch.GetConnection();
    }

    private async Task SleepAsync(int milliseconds)
    {
        await Task.Delay(milliseconds);
    }

    public async Task<dynamic> FindRpcClientIdAsync(
        string? kb = null,
        string? nodeName = null,
        Dictionary<string, object>? properties = null,
        string? nodePath = null)
    {
        var ids = await FindRpcClientIdsAsync(kb, nodeName, properties, nodePath);
        
        if (ids.Count == 0)
        {
            throw new InvalidOperationException(
                $"No node found matching parameters: {nodeName}, {JsonConvert.SerializeObject(properties)}, {nodePath}");
        }
        
        if (ids.Count > 1)
        {
            throw new InvalidOperationException(
                $"Multiple nodes found matching parameters: {nodeName}, {JsonConvert.SerializeObject(properties)}, {nodePath}");
        }
        
        return ids[0];
    }

    public async Task<List<dynamic>> FindRpcClientIdsAsync(
        string? kb = null,
        string? nodeName = null,
        Dictionary<string, object>? properties = null,
        string? nodePath = null)
    {
        _kbSearch.ClearFilters();
        _kbSearch.SearchLabel("KB_RPC_CLIENT_FIELD");
        
        if (!string.IsNullOrEmpty(kb)) _kbSearch.SearchKb(kb);
        if (!string.IsNullOrEmpty(nodeName)) _kbSearch.SearchName(nodeName);
        
        if (properties != null)
        {
            foreach (var key in properties.Keys)
            {
                _kbSearch.SearchPropertyValue(key, properties[key]);
            }
        }
        
        if (!string.IsNullOrEmpty(nodePath)) _kbSearch.SearchPath(nodePath);

        var results = await _kbSearch.ExecuteQueryAsync();
        
        if (results == null || results.Count == 0)
        {
            throw new InvalidOperationException(
                $"No node found matching parameters: {nodeName}, {JsonConvert.SerializeObject(properties)}, {nodePath}");
        }
        
        return results;
    }

    public List<string> FindRpcClientKeys(List<dynamic> keyData)
    {
        return keyData.Select(r => ((Dictionary<string, object>)r)["path"]?.ToString() ?? string.Empty).ToList();
    }

    public async Task<int> FindFreeSlotsAsync(string clientPath)
    {
        var client = GetConnection();
        var query = $@"
            SELECT COUNT(*) AS total_records,
                   COUNT(*) FILTER (WHERE is_new_result = FALSE) AS free_slots
            FROM {_baseTable}
            WHERE client_path = $1::ltree
        ";
        
        try
        {
            using var cmd = new NpgsqlCommand(query, client);
            cmd.Parameters.AddWithValue(clientPath);
            
            using var reader = await cmd.ExecuteReaderAsync();
            await reader.ReadAsync();
            
            var total = Convert.ToInt32(reader["total_records"]);
            var free = Convert.ToInt32(reader["free_slots"]);
            
            if (total == 0)
            {
                throw new InvalidOperationException($"No records found for client_path: {clientPath}");
            }
            
            return free;
        }
        catch (Exception ex) when (!(ex is InvalidOperationException))
        {
            throw new Exception($"Database error when finding free slots: {ex.Message}", ex);
        }
    }

    public async Task<int> FindQueuedSlotsAsync(string clientPath)
    {
        var client = GetConnection();
        var query = $@"
            SELECT COUNT(*) AS total_records,
                   COUNT(*) FILTER (WHERE is_new_result = TRUE) AS queued_slots
            FROM {_baseTable}
            WHERE client_path = $1::ltree
        ";
        
        try
        {
            using var cmd = new NpgsqlCommand(query, client);
            cmd.Parameters.AddWithValue(clientPath);
            
            using var reader = await cmd.ExecuteReaderAsync();
            await reader.ReadAsync();
            
            var total = Convert.ToInt32(reader["total_records"]);
            var queued = Convert.ToInt32(reader["queued_slots"]);
            
            if (total == 0)
            {
                throw new InvalidOperationException($"No records found for client_path: {clientPath}");
            }
            
            return queued;
        }
        catch (Exception ex) when (!(ex is InvalidOperationException))
        {
            throw new Exception($"Database error when finding queued slots: {ex.Message}", ex);
        }
    }

    public async Task<Dictionary<string, object>?> PeakAndClaimReplyDataAsync(
        string clientPath,
        int maxRetries = 3,
        int retryDelay = 1000)
    {
        var client = GetConnection();
        int attempt = 0;
        var table = _baseTable;
        
        while (attempt < maxRetries)
        {
            try
            {
                using var transaction = await client.BeginTransactionAsync();
                
                try
                {
                    var updateQuery = $@"
                        UPDATE {table}
                        SET is_new_result = FALSE
                        WHERE id = (
                            SELECT id FROM {table}
                            WHERE client_path = $1::ltree AND is_new_result = TRUE
                            ORDER BY response_timestamp ASC
                            FOR UPDATE SKIP LOCKED LIMIT 1
                        )
                        RETURNING *
                    ";
                    
                    using var updateCmd = new NpgsqlCommand(updateQuery, client, transaction);
                    updateCmd.Parameters.AddWithValue(clientPath);
                    
                    using var updateReader = await updateCmd.ExecuteReaderAsync();
                    
                    if (await updateReader.ReadAsync())
                    {
                        var result = new Dictionary<string, object>();
                        for (int i = 0; i < updateReader.FieldCount; i++)
                        {
                            result[updateReader.GetName(i)] = updateReader.IsDBNull(i) ? null : updateReader.GetValue(i);
                        }
                        
                        await updateReader.CloseAsync();
                        await transaction.CommitAsync();
                        return result;
                    }
                    
                    await updateReader.CloseAsync();
                    
                    // Check existence
                    var existQuery = $@"
                        SELECT EXISTS(
                            SELECT 1 FROM {table}
                            WHERE client_path = $1::ltree AND is_new_result = TRUE
                        ) AS exists
                    ";
                    
                    using var existCmd = new NpgsqlCommand(existQuery, client, transaction);
                    existCmd.Parameters.AddWithValue(clientPath);
                    
                    var exists = (bool)await existCmd.ExecuteScalarAsync();
                    
                    if (!exists)
                    {
                        await transaction.RollbackAsync();
                        return null;
                    }
                    
                    await transaction.RollbackAsync();
                    attempt++;
                    await SleepAsync(retryDelay);
                }
                catch
                {
                    await transaction.RollbackAsync();
                    throw;
                }
            }
            catch (Exception)
            {
                attempt++;
                await SleepAsync(retryDelay);
            }
        }
        
        throw new Exception($"Could not lock a new-reply row after {maxRetries} attempts");
    }

    public async Task<int> ClearReplyQueueAsync(
        string clientPath,
        int maxRetries = 3,
        int retryDelay = 1000)
    {
        var client = GetConnection();
        var table = _baseTable;
        
        // First check if any records exist for this client path
        var countQuery = $@"
            SELECT COUNT(*) AS count
            FROM {table}
            WHERE client_path = $1::ltree
        ";
        
        using var countCmd = new NpgsqlCommand(countQuery, client);
        countCmd.Parameters.AddWithValue(clientPath);
        var recordCount = Convert.ToInt32(await countCmd.ExecuteScalarAsync());
        
        if (recordCount == 0)
        {
            // No records exist for this client path, so nothing to clear
            return 0;
        }
        
        int attempt = 0;
        
        while (attempt < maxRetries)
        {
            try
            {
                using var transaction = await client.BeginTransactionAsync();
                
                try
                {
                    var selectQuery = $@"
                        SELECT id FROM {table}
                        WHERE client_path = $1::ltree FOR UPDATE NOWAIT
                    ";
                    
                    using var selectCmd = new NpgsqlCommand(selectQuery, client, transaction);
                    selectCmd.Parameters.AddWithValue(clientPath);
                    
                    using var selectReader = await selectCmd.ExecuteReaderAsync();
                    var ids = new List<int>();
                    
                    while (await selectReader.ReadAsync())
                    {
                        var idValue = selectReader["id"];
                        ids.Add(Convert.ToInt32(idValue));
                    }
                    
                    await selectReader.CloseAsync();
                    
                    if (ids.Count == 0)
                    {
                        await transaction.CommitAsync();
                        return 0;
                    }
                    
                    int updated = 0;
                    foreach (var id in ids)
                    {
                        var updateQuery = $@"
                            UPDATE {table}
                            SET request_id = $1::uuid,
                                server_path = $2::ltree,
                                response_payload = $3::json,
                                response_timestamp = NOW(),
                                is_new_result = FALSE
                            WHERE id = $4
                        ";
                        
                        using var updateCmd = new NpgsqlCommand(updateQuery, client, transaction);
                        var uuid = Guid.NewGuid().ToString();
                        updateCmd.Parameters.AddWithValue(uuid);
                        updateCmd.Parameters.AddWithValue(clientPath);
                        updateCmd.Parameters.AddWithValue(JsonConvert.SerializeObject(new { }));
                        updateCmd.Parameters.AddWithValue(id);
                        
                        await updateCmd.ExecuteNonQueryAsync();
                        updated++;
                    }
                    
                    await transaction.CommitAsync();
                    return updated;
                }
                catch
                {
                    await transaction.RollbackAsync();
                    throw;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Clear reply queue attempt {attempt + 1} failed: {ex.Message}");
                attempt++;
                if (attempt < maxRetries)
                {
                    await SleepAsync(retryDelay);
                }
            }
        }
        
        throw new Exception($"Could not acquire lock after {maxRetries} retries");
    }

    public async Task PushAndClaimReplyDataAsync(
        string clientPath,
        string requestUuid,
        string serverPath,
        string rpcAction,
        string transactionTag,
        object replyData,
        int maxRetries = 3,
        int retryDelay = 1000)
    {
        var client = GetConnection();
        int attempt = 0;
        var table = _baseTable;
        
        while (attempt <= maxRetries)
        {
            try
            {
                using var transaction = await client.BeginTransactionAsync();
                
                try
                {
                    var upsertQuery = $@"
                        WITH candidate AS (
                            SELECT id FROM {table}
                            WHERE client_path = $1::ltree AND is_new_result = FALSE
                            ORDER BY response_timestamp ASC
                            FOR UPDATE SKIP LOCKED LIMIT 1
                        )
                        UPDATE {table}
                        SET request_id = $2::uuid, server_path = $3::ltree, rpc_action = $4,
                            transaction_tag = $5, response_payload = $6::json,
                            is_new_result = TRUE, response_timestamp = CURRENT_TIMESTAMP
                        FROM candidate
                        WHERE {table}.id = candidate.id
                        RETURNING {table}.id;
                    ";
                    
                    using var cmd = new NpgsqlCommand(upsertQuery, client, transaction);
                    cmd.Parameters.AddWithValue(clientPath);
                    cmd.Parameters.AddWithValue(requestUuid);
                    cmd.Parameters.AddWithValue(serverPath);
                    cmd.Parameters.AddWithValue(rpcAction);
                    cmd.Parameters.AddWithValue(transactionTag);
                    cmd.Parameters.AddWithValue(JsonConvert.SerializeObject(replyData));
                    
                    var result = await cmd.ExecuteScalarAsync();
                    
                    if (result == null)
                    {
                        await transaction.RollbackAsync();
                        throw new InvalidOperationException("No available record found");
                    }
                    
                    await transaction.CommitAsync();
                    return;
                }
                catch
                {
                    await transaction.RollbackAsync();
                    throw;
                }
            }
            catch (Exception )
            {
                attempt++;
                if (attempt > maxRetries) throw;
                await SleepAsync(retryDelay);
            }
        }
    }

    public async Task<List<Dictionary<string, object>>> ListWaitingJobsAsync(string? clientPath = null)
    {
        var client = GetConnection();
        var table = _baseTable;
        var query = $@"
            SELECT id, request_id, client_path, server_path,
                   response_payload, response_timestamp, is_new_result
            FROM {table}
            WHERE is_new_result = TRUE
        ";
        
        var parameters = new List<object>();
        
        if (!string.IsNullOrEmpty(clientPath))
        {
            query += " AND client_path = $1::ltree";
            parameters.Add(clientPath);
        }
        
        query += " ORDER BY response_timestamp ASC";

        try
        {
            using var cmd = new NpgsqlCommand(query, client);
            for (int i = 0; i < parameters.Count; i++)
            {
                cmd.Parameters.AddWithValue(parameters[i]);
            }
            
            using var reader = await cmd.ExecuteReaderAsync();
            var results = new List<Dictionary<string, object>>();
            
            while (await reader.ReadAsync())
            {
                var row = new Dictionary<string, object>();
                
                for (int i = 0; i < reader.FieldCount; i++)
                {
                    var columnName = reader.GetName(i);
                    var value = reader.IsDBNull(i) ? null! : reader.GetValue(i)!;
                    
                    // Handle special transformations
                    if (columnName == "request_id" && value != null)
                    {
                        value = value.ToString();
                    }
                    else if (columnName == "response_timestamp" && value is DateTime dateTime)
                    {
                        value = dateTime.ToString("yyyy-MM-ddTHH:mm:ss.fffZ");
                    }
                    
                    row[columnName] = value;
                }
                
                results.Add(row);
            }
            
            return results;
        }
        catch (Exception ex)
        {
            throw new Exception($"Database error when listing waiting jobs: {ex.Message}", ex);
        }
    }
}